Fix virt.get_profiles

rst0git opened the pull request:
### What does this PR do?
This PR fixes a bug in virt.get_profiles and updates the integration tests.

### Previous Behavior
```
# salt guest virt.get_profiles hypervisor=kvm
guest:
    Passed invalid arguments to virt.get_profiles: _disk_profile() missing 3 required positional arguments: 'hypervisor', 'disks', and 'vm_name'
    
        Return the virt profiles for hypervisor.
    
        Currently there are profiles for:
    
        - nic
        - disk
    
        :param hypervisor: override the default machine type.
        :param connection: libvirt connection URI, overriding defaults
    
            .. versionadded:: 2019.2.0
        :param username: username to connect with, overriding defaults
    
            .. versionadded:: 2019.2.0
        :param password: password to connect with, overriding defaults
    
            .. versionadded:: 2019.2.0
    
        CLI Example:
    
        .. code-block:: bash
    
            salt '*' virt.get_profiles
            salt '*' virt.get_profiles hypervisor=esxi
        
ERROR: Minions returned with non-zero exit code
```

### New Behavior
```
# salt guest virt.get_profiles hypervisor=kvm
guest:
    ----------
    disk:
        ----------
        default:
            |_
              ----------
              device:
                  disk
              model:
                  virtio
              name:
                  system
              size:
                  8192
    nic:
        ----------
        default:
            |_
              ----------
              model:
                  virtio
              name:
                  eth0
              source:
                  br0
              type:
                  bridge
```

rst0git committed:
virt: Remove unused imports

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
integration/test_virt: Assert 'profiles' type

When get_profiles() fails, it returns a string with an error message.
This causes the integration tests to fail as follows:

-> integration.modules.test_virt.VirtTest.test_default_esxi_profile
	Traceback (most recent call last):
	File "/salt/tests/integration/modules/test_virt.py", line 38, in test_default_esxi_profile
		nicp = profiles["nic"]["default"]
	TypeError: string indices must be integers

By asserting the return type of get_profiles(), the error message would
indicate the actual reason for the failure:

-> integration.modules.test_virt.VirtTest.test_default_esxi_profile
	Traceback (most recent call last):
	File "/salt/tests/integration/modules/test_virt.py", line 39, in test_default_esxi_profile
		self.assertIsInstance(profiles, dict)
	AssertionError: 'ERROR: Sorry, localhost failed to open a connection to the hypervisor software at None' is not an instance of <class 'dict'>

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
virt: Handle no available hypervisors

The output from libvirt's virConnectGetCapabilities
might contain zero or more `guest` elements [1].

This commit introduces an appropriate error message shown
when the number of guest elements is 0.

[1] https://libvirt.org/formatcaps.html

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
virt: Refactor get_profiles() and integration tests

In commit [1] the use of the 'esxi' hypervisor was replaced with
'vmware'. This commit updates the integration test and documentation
string to match this change.

A number of new arguments have been introduced to _disk_profile(),
however, get_profiles() assumes that both _disk_profile() and
_nic_profile() take as input the same arguments.

In particular, commit [2] introduced the `pool` argument and
commit [3] introduced `disks`, `vm_name` and `image`. In commit [4]
were removed the default values for the new arguments added to
_disk_profile(), which results in get_profiles() to fail with
'invalid arguments' error. The changes introduced in this commit
should resolve this error. In addition, the integration tests have
been updated to reflect this and other changes.

The integration tests require the Python bindings for libvirt to be
installed (`pip install libvirt-python`) and a local libvirt daemon
instance (`systemctl start libvirtd`).

The tests can be executed with:

	python tests/runtests.py -n integration.modules.test_virt

[1] https://github.com/saltstack/salt/commit/34d1b34
[2] https://github.com/saltstack/salt/commit/7555172
[3] https://github.com/saltstack/salt/commit/5479f88
[4] https://github.com/saltstack/salt/commit/8ed54cd

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git commented:
cc @cbosdo

cbosdo reviewed (dismissed):
Thanks for the fixes. For the integration tests we will need to have @waynew or @Ch3LL feedback since now it requires libvirt to be running.

cbosdo reviewed (dismissed):
:+1:

rst0git commented:
> For the integration tests we will need to have @waynew or @Ch3LL feedback since now it requires libvirt to be running.

Some tests we might need more than one libvirt daemon instance (e.g. live migration). The CentOS 7 test VM already has docker installed as a dependency for the Docker integration tests.
We could use docker to run multiple libvirt + salt minion containers for the virt integration tests.

waynew commented:
The Docker approach seems reasonable to me. I'm not intimately familiar with virt, though, so discussion is definitely welcome 👍 

FYI these Windows failures should be fixed with the latest `master` if you want to rebase or merge.

waynew commented:
re-run centos7-py3 centos7-py3-cryptodome

waynew commented:
:green_heart:

waynew commented:
Looks like existing tests are good. Do we still need to add some other tests for this?

cbosdo reviewed (dismissed):
:+1:

rst0git commented:
> Do we still need to add some other tests for this?

I will add more integration tests for the virt module in another PR.

rst0git committed:
virt: Don't expose check_remote and download_remote

The bash autocomplete includes the utility functions check_remote and
download_remote. These are functions are not intended to be directly
used by users.

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
Add virt.all_capabilities integration test

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
Add virt.capabilities integration test

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
Add virt.cpu_baseline integration test

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
Add virt freecpu/freemem integration test

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git committed:
Add virt full_info/node_info integration tests

Signed-off-by: Radostin Stoyanov <rstoyanov@fedoraproject.org>

rst0git commented:
I rebased the PR to resolve a conflict with the latest `master` and added a few more integration tests.

waynew reviewed (approved):
:rocket:

rst0git commented:
Closing in favour of https://github.com/saltstack/salt/pull/57947
